{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <font color = red> Project: \n",
    "## Predict Kidney Tumor Stage With Machine Learning Using Epigenomic Data "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <font color = red> Problem: \n",
    "## High misdiagnosis rate of tumor stage, which is important for doctors to choose suitable therapy for patients"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <font color = red> Data Pipeline:\n",
    "## 1. Download data sets from TCGA database: (i). Patients' clinical diagnosis data (ii). DNA          methylation data\n",
    "## 2. Cleaning and merging two datasets\n",
    "## 3. Feature Selection: (i). PCA (ii). LDA (iii). Kruskal-Wallis Method\n",
    "## 4. Classifier: (i). LDA (ii). Logistic Regression (iii). Random Forest "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color = red> Results:\n",
    "### 1. Tested several different combinations of dimension reduction methods (PCA, LDA, Kruskal- Wallis test) and classifiers (Logistic regression, LDA, Random forest) in Python; combinations of Kruskal-Wallis test and random forest classifier have highest accuracy in predicting tumor stage \n",
    "### 2. Analyzed 208 patients with over 20,000 genes methylation data (from TCGA database) and identified 20 genes as key features for tumor stage prediction; accuracy of prediction is over 70%"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 1. Data Loading, Cleaning, and Merge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Import necessary libraries\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.1 Clinical Diagnosis Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_barcode</th>\n",
       "      <th>pathologic_stage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tcga-a3-3307</td>\n",
       "      <td>stage iii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tcga-a3-3325</td>\n",
       "      <td>stage i</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tcga-a3-3328</td>\n",
       "      <td>stage i</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tcga-a3-3329</td>\n",
       "      <td>stage i</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tcga-a3-3336</td>\n",
       "      <td>stage i</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  patient_barcode pathologic_stage\n",
       "0    tcga-a3-3307        stage iii\n",
       "1    tcga-a3-3325          stage i\n",
       "2    tcga-a3-3328          stage i\n",
       "3    tcga-a3-3329          stage i\n",
       "4    tcga-a3-3336          stage i"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## First dataset: Patients' clinical diagnosis data \n",
    "data1 = pd.read_csv('KIRC.clin.merged.picked.txt', delimiter=\"\\t\")\n",
    "\n",
    "## Get pathological stage data\n",
    "clinical = data1.iloc[6:7,:]\n",
    "clinical = clinical.rename(columns = {'Hybridization REF':'Patient_Barcode'}) # df = df.rename(columns={'oldName1': 'newName1', 'oldName2': 'newName2'})\n",
    "\n",
    "## Transpose data & rename\n",
    "B = clinical.transpose()\n",
    "B.head()\n",
    "\n",
    "# B.columns = [\"pathologic_stage\"]\n",
    "clinical_final = B.iloc[1:,:]\n",
    "clinical_final=clinical_final.reset_index()\n",
    "clinical_final.columns = [\"patient_barcode\",\"pathologic_stage\"]\n",
    "clinical_final.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.2 DNA Methylation Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/czl/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3020: DtypeWarning: Columns (1,4,5,8,9,12,13,16,17,20,21,24,25,28,29,32,33,36,37,40,41,44,45,48,49,52,53,56,57,60,61,64,65,68,69,72,73,76,77,80,81,84,85,88,89,92,93,96,97,100,101,104,105,108,109,112,113,116,117,120,121,124,125,128,129,132,133,136,137,140,141,144,145,148,149,152,153,156,157,160,161,164,165,168,169,172,173,176,177,180,181,184,185,188,189,192,193,196,197,200,201,204,205,208,209,212,213,216,217,220,221,224,225,228,229,232,233,236,237,240,241,244,245,248,249,252,253,256,257,260,261,264,265,268,269,272,273,276,277,280,281,284,285,288,289,292,293,296,297,300,301,304,305,308,309,312,313,316,317,320,321,324,325,328,329,332,333,336,337,340,341,344,345,348,349,352,353,356,357,360,361,364,365,368,369,372,373,376,377,380,381,384,385,388,389,392,393,396,397,400,401,404,405,408,409,412,413,416,417,420,421,424,425,428,429,432,433,436,437,440,441,444,445,448,449,452,453,456,457,460,461,464,465,468,469,472,473,476,477,480,481,484,485,488,489,492,493,496,497,500,501,504,505,508,509,512,513,516,517,520,521,524,525,528,529,532,533,536,537,540,541,544,545,548,549,552,553,556,557,560,561,564,565,568,569,572,573,576,577,580,581,584,585,588,589,592,593,596,597,600,601,604,605,608,609,612,613,616,617,620,621,624,625,628,629,632,633,636,637,640,641,644,645,648,649,652,653,656,657,660,661,664,665,668,669,672,673,676,677,680,681,684,685,688,689,692,693,696,697,700,701,704,705,708,709,712,713,716,717,720,721,724,725,728,729,732,733,736,737,740,741,744,745,748,749,752,753,756,757,760,761,764,765,768,769,772,773,776,777,780,781,784,785,788,789,792,793,796,797,800,801,804,805,808,809,812,813,816,817,820,821,824,825,828,829,832,833,836,837,840,841,844,845,848,849,852,853,856,857,860,861,864,865,868,869,872,873,876,877,880,881,884,885,888,889,892,893,896,897,900,901,904,905,908,909,912,913,916,917,920,921,924,925,928,929,932,933,936,937,940,941,944,945,948,949,952,953,956,957,960,961,964,965,968,969,972,973,976,977,980,981,984,985,988,989,992,993,996,997,1000,1001,1004,1005,1008,1009,1012,1013,1016,1017,1020,1021,1024,1025,1028,1029,1032,1033,1036,1037,1040,1041,1044,1045,1048,1049,1052,1053,1056,1057,1060,1061,1064,1065,1068,1069,1072,1073,1076,1077,1080,1081,1084,1085,1088,1089,1092,1093,1096,1097,1100,1101,1104,1105,1108,1109,1112,1113,1116,1117,1120,1121,1124,1125,1128,1129,1132,1133,1136,1137,1140,1141,1144,1145,1148,1149,1152,1153,1156,1157,1160,1161,1164,1165,1168,1169,1172,1173,1176,1177,1180,1181,1184,1185,1188,1189,1192,1193,1196,1197,1200,1201,1204,1205,1208,1209,1212,1213,1216,1217,1220,1221,1224,1225,1228,1229,1232,1233,1236,1237,1240,1241,1244,1245,1248,1249,1252,1253,1256,1257,1260,1261,1264,1265,1268,1269,1272,1273,1276,1277,1280,1281,1284,1285,1288,1289,1292,1293,1296,1297,1300,1301,1304,1305,1308,1309,1312,1313,1316,1317,1320,1321,1324,1325,1328,1329,1332,1333,1336,1337,1340,1341,1344,1345,1348,1349,1352,1353,1356,1357,1360,1361,1364,1365,1368,1369,1372,1373,1376,1377,1380,1381,1384,1385,1388,1389,1392,1393,1396,1397,1400,1401,1404,1405,1408,1409,1412,1413,1416,1417,1420,1421,1424,1425,1428,1429,1432,1433,1436,1437,1440,1441,1444,1445,1448,1449,1452,1453,1456,1457,1460,1461,1464,1465,1468,1469,1472,1473,1476,1477,1480,1481,1484,1485,1488,1489,1492,1493,1496,1497,1500,1501,1504,1505,1508,1509,1512,1513,1516,1517,1520,1521,1524,1525,1528,1529,1532,1533,1536,1537,1540,1541,1544,1545,1548,1549,1552,1553,1556,1557,1560,1561,1564,1565,1568,1569,1572,1573,1576,1577,1580,1581,1584,1585,1588,1589,1592,1593,1596,1597,1600,1601,1604,1605,1608,1609,1612,1613,1616,1617,1620,1621,1624,1625,1628,1629,1632,1633,1636,1637,1640,1641,1644,1645,1648,1649,1652,1653,1656,1657,1660,1661,1664,1665,1668,1669,1672) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "## Second dataset: DNA Methylation Data\n",
    "url = 'KIRC.methylation__humanmethylation27__jhu_usc_edu__Level_3__within_bioassay_data_set_function__data.data.txt'\n",
    "meth = pd.read_csv(url, delimiter=\"\\t\")\n",
    "## Choose beta-value & gene symbol\n",
    "meth = meth.transpose().reset_index()\n",
    "meth = meth.iloc[1:,:]\n",
    "\n",
    "## Rename column with gene names\n",
    "header = ['patient','category']+list(meth.iloc[1,2:])\n",
    "\n",
    "## Mark different locations in the same gene as: geneA and geneA_1\n",
    "# Ex: ATP2A1 and ATP2A1_1\n",
    "m = set()\n",
    "for n in range(len(header)):\n",
    "    if str(header[n]) in m:\n",
    "        header[n] = str(header[n])+'_1'\n",
    "        m.add(header[n])\n",
    "    else:\n",
    "        m.add(header[n])\n",
    "meth = meth.set_axis(header, axis='columns', inplace=False)\n",
    "\n",
    "## Drop rows with 'Chromosome' & 'Genomic Coordinate' & 'Gene_Symbol' item in 'category' column\n",
    "meth = meth[~meth.category.str.contains('Chromosome')]\n",
    "meth = meth[~meth.category.str.contains('Genomic_Coordinate')]\n",
    "meth = meth[~meth.category.str.contains('Gene_Symbol')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.3 Merge Datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_barcode</th>\n",
       "      <th>pathologic_stage</th>\n",
       "      <th>patient</th>\n",
       "      <th>category</th>\n",
       "      <th>ATP2A1</th>\n",
       "      <th>SLMAP</th>\n",
       "      <th>MEOX2</th>\n",
       "      <th>HOXD3</th>\n",
       "      <th>ZNF425;ZNF398</th>\n",
       "      <th>PANX1</th>\n",
       "      <th>...</th>\n",
       "      <th>A2ML1_1</th>\n",
       "      <th>ZFP2_1</th>\n",
       "      <th>CST9L_1</th>\n",
       "      <th>C11orf24_1</th>\n",
       "      <th>nan</th>\n",
       "      <th>LEPRE1;C1orf50_1</th>\n",
       "      <th>GNAS_1</th>\n",
       "      <th>RPN1_1</th>\n",
       "      <th>CYB5A_1</th>\n",
       "      <th>AP1S1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tcga-a3-3325</td>\n",
       "      <td>stage i</td>\n",
       "      <td>TCGA-A3-3325-11A-01D-0859-05</td>\n",
       "      <td>Beta_value</td>\n",
       "      <td>0.341092896642809</td>\n",
       "      <td>0.163420373331899</td>\n",
       "      <td>0.0289863381040888</td>\n",
       "      <td>0.704490559626231</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0393644321967782</td>\n",
       "      <td>...</td>\n",
       "      <td>0.800048</td>\n",
       "      <td>0.0221649</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0195139</td>\n",
       "      <td>0.18193</td>\n",
       "      <td>0.0284293</td>\n",
       "      <td>0.389761</td>\n",
       "      <td>0.0101311</td>\n",
       "      <td>0.0167906</td>\n",
       "      <td>0.0393153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tcga-a3-3325</td>\n",
       "      <td>stage i</td>\n",
       "      <td>TCGA-A3-3325-01A-01D-0859-05</td>\n",
       "      <td>Beta_value</td>\n",
       "      <td>0.7253328141359</td>\n",
       "      <td>0.235892179963932</td>\n",
       "      <td>0.237244972745325</td>\n",
       "      <td>0.813347696270696</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0513241779156893</td>\n",
       "      <td>...</td>\n",
       "      <td>0.530865</td>\n",
       "      <td>0.0235912</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0211279</td>\n",
       "      <td>0.186888</td>\n",
       "      <td>0.0288994</td>\n",
       "      <td>0.371611</td>\n",
       "      <td>0.0100788</td>\n",
       "      <td>0.0142257</td>\n",
       "      <td>0.0132333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tcga-a3-3328</td>\n",
       "      <td>stage i</td>\n",
       "      <td>TCGA-A3-3328-11A-01D-0859-05</td>\n",
       "      <td>Beta_value</td>\n",
       "      <td>0.38595358991842</td>\n",
       "      <td>0.172516336695289</td>\n",
       "      <td>0.0267564273359283</td>\n",
       "      <td>0.737048613855687</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.032547823619346</td>\n",
       "      <td>...</td>\n",
       "      <td>0.800673</td>\n",
       "      <td>0.017531</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0177181</td>\n",
       "      <td>0.185258</td>\n",
       "      <td>0.0282834</td>\n",
       "      <td>0.363502</td>\n",
       "      <td>0.0098538</td>\n",
       "      <td>0.0156143</td>\n",
       "      <td>0.0360442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tcga-a3-3328</td>\n",
       "      <td>stage i</td>\n",
       "      <td>TCGA-A3-3328-01A-01D-0859-05</td>\n",
       "      <td>Beta_value</td>\n",
       "      <td>0.88549631227471</td>\n",
       "      <td>0.0384832173567634</td>\n",
       "      <td>0.0235068147586301</td>\n",
       "      <td>0.109113302762268</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0486053663743769</td>\n",
       "      <td>...</td>\n",
       "      <td>0.903917</td>\n",
       "      <td>0.017625</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0189847</td>\n",
       "      <td>0.179154</td>\n",
       "      <td>0.05492</td>\n",
       "      <td>0.377929</td>\n",
       "      <td>0.00975378</td>\n",
       "      <td>0.0150563</td>\n",
       "      <td>0.0146956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tcga-a3-3329</td>\n",
       "      <td>stage i</td>\n",
       "      <td>TCGA-A3-3329-11A-01D-0859-05</td>\n",
       "      <td>Beta_value</td>\n",
       "      <td>0.257540543081615</td>\n",
       "      <td>0.126193684266256</td>\n",
       "      <td>0.0453059118170011</td>\n",
       "      <td>0.658352096657674</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0417558123692779</td>\n",
       "      <td>...</td>\n",
       "      <td>0.786991</td>\n",
       "      <td>0.0213796</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0199632</td>\n",
       "      <td>0.160584</td>\n",
       "      <td>0.046203</td>\n",
       "      <td>0.337231</td>\n",
       "      <td>0.0105036</td>\n",
       "      <td>0.017995</td>\n",
       "      <td>0.0418536</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 27582 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  patient_barcode pathologic_stage                       patient    category  \\\n",
       "0    tcga-a3-3325          stage i  TCGA-A3-3325-11A-01D-0859-05  Beta_value   \n",
       "1    tcga-a3-3325          stage i  TCGA-A3-3325-01A-01D-0859-05  Beta_value   \n",
       "2    tcga-a3-3328          stage i  TCGA-A3-3328-11A-01D-0859-05  Beta_value   \n",
       "3    tcga-a3-3328          stage i  TCGA-A3-3328-01A-01D-0859-05  Beta_value   \n",
       "4    tcga-a3-3329          stage i  TCGA-A3-3329-11A-01D-0859-05  Beta_value   \n",
       "\n",
       "              ATP2A1               SLMAP               MEOX2  \\\n",
       "0  0.341092896642809   0.163420373331899  0.0289863381040888   \n",
       "1    0.7253328141359   0.235892179963932   0.237244972745325   \n",
       "2   0.38595358991842   0.172516336695289  0.0267564273359283   \n",
       "3   0.88549631227471  0.0384832173567634  0.0235068147586301   \n",
       "4  0.257540543081615   0.126193684266256  0.0453059118170011   \n",
       "\n",
       "               HOXD3 ZNF425;ZNF398               PANX1    ...       A2ML1_1  \\\n",
       "0  0.704490559626231           NaN  0.0393644321967782    ...      0.800048   \n",
       "1  0.813347696270696           NaN  0.0513241779156893    ...      0.530865   \n",
       "2  0.737048613855687           NaN   0.032547823619346    ...      0.800673   \n",
       "3  0.109113302762268           NaN  0.0486053663743769    ...      0.903917   \n",
       "4  0.658352096657674           NaN  0.0417558123692779    ...      0.786991   \n",
       "\n",
       "      ZFP2_1 CST9L_1 C11orf24_1       NaN LEPRE1;C1orf50_1    GNAS_1  \\\n",
       "0  0.0221649     NaN  0.0195139   0.18193        0.0284293  0.389761   \n",
       "1  0.0235912     NaN  0.0211279  0.186888        0.0288994  0.371611   \n",
       "2   0.017531     NaN  0.0177181  0.185258        0.0282834  0.363502   \n",
       "3   0.017625     NaN  0.0189847  0.179154          0.05492  0.377929   \n",
       "4  0.0213796     NaN  0.0199632  0.160584         0.046203  0.337231   \n",
       "\n",
       "       RPN1_1    CYB5A_1      AP1S1  \n",
       "0   0.0101311  0.0167906  0.0393153  \n",
       "1   0.0100788  0.0142257  0.0132333  \n",
       "2   0.0098538  0.0156143  0.0360442  \n",
       "3  0.00975378  0.0150563  0.0146956  \n",
       "4   0.0105036   0.017995  0.0418536  \n",
       "\n",
       "[5 rows x 27582 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Merge two datasets based on patient ID\n",
    "    # 1. Create a new column named column, which contains first 12 letters of patient_code\n",
    "meth['col'] = meth['patient'].str[0:12]\n",
    "meth['col_lower'] = meth['col'].str.lower() # Convert 'col' column to lower case\n",
    "meth_final = meth.rename(columns = {'col_lower':'patient_barcode'})\n",
    "\n",
    "# patient: TCGA-A3-3306-11A (tumor sample) v.s. TCGA-A3-3306-01A (normal sample)\n",
    "# https://docs.gdc.cancer.gov/Encyclopedia/pages/TCGA_Barcode/\n",
    "\n",
    "    # 2.Merge clinical data & DNA methylation data\n",
    "mixed_df = pd.merge(clinical_final, meth_final, how='inner', on=['patient_barcode']) # Choose overlapped region\n",
    "mixed_df = mixed_df.drop(['col'],axis=1)\n",
    "mixed_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/czl/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:3140: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  self[k1] = value[k2]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_barcode</th>\n",
       "      <th>patient</th>\n",
       "      <th>category</th>\n",
       "      <th>ATP2A1</th>\n",
       "      <th>SLMAP</th>\n",
       "      <th>MEOX2</th>\n",
       "      <th>HOXD3</th>\n",
       "      <th>PANX1</th>\n",
       "      <th>COX8C;KIAA1409</th>\n",
       "      <th>TTC8</th>\n",
       "      <th>...</th>\n",
       "      <th>MYCL1_1</th>\n",
       "      <th>KIAA1409;COX8C</th>\n",
       "      <th>ALDH1A3_1</th>\n",
       "      <th>ZFP2_1</th>\n",
       "      <th>C11orf24_1</th>\n",
       "      <th>LEPRE1;C1orf50_1</th>\n",
       "      <th>GNAS_1</th>\n",
       "      <th>RPN1_1</th>\n",
       "      <th>CYB5A_1</th>\n",
       "      <th>pathologic_stage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tcga-a3-3325</td>\n",
       "      <td>TCGA-A3-3325-11A-01D-0859-05</td>\n",
       "      <td>Beta_value</td>\n",
       "      <td>0.341092896642809</td>\n",
       "      <td>0.163420373331899</td>\n",
       "      <td>0.0289863381040888</td>\n",
       "      <td>0.704490559626231</td>\n",
       "      <td>0.0393644321967782</td>\n",
       "      <td>0.988790175005561</td>\n",
       "      <td>0.0101987967341247</td>\n",
       "      <td>...</td>\n",
       "      <td>0.034498</td>\n",
       "      <td>0.982875</td>\n",
       "      <td>0.280882</td>\n",
       "      <td>0.0221649</td>\n",
       "      <td>0.0195139</td>\n",
       "      <td>0.0284293</td>\n",
       "      <td>0.389761</td>\n",
       "      <td>0.0101311</td>\n",
       "      <td>0.0167906</td>\n",
       "      <td>stage i</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tcga-a3-3325</td>\n",
       "      <td>TCGA-A3-3325-01A-01D-0859-05</td>\n",
       "      <td>Beta_value</td>\n",
       "      <td>0.7253328141359</td>\n",
       "      <td>0.235892179963932</td>\n",
       "      <td>0.237244972745325</td>\n",
       "      <td>0.813347696270696</td>\n",
       "      <td>0.0513241779156893</td>\n",
       "      <td>0.988732478437813</td>\n",
       "      <td>0.00849065030592262</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0397872</td>\n",
       "      <td>0.97624</td>\n",
       "      <td>0.471156</td>\n",
       "      <td>0.0235912</td>\n",
       "      <td>0.0211279</td>\n",
       "      <td>0.0288994</td>\n",
       "      <td>0.371611</td>\n",
       "      <td>0.0100788</td>\n",
       "      <td>0.0142257</td>\n",
       "      <td>stage i</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tcga-a3-3328</td>\n",
       "      <td>TCGA-A3-3328-11A-01D-0859-05</td>\n",
       "      <td>Beta_value</td>\n",
       "      <td>0.38595358991842</td>\n",
       "      <td>0.172516336695289</td>\n",
       "      <td>0.0267564273359283</td>\n",
       "      <td>0.737048613855687</td>\n",
       "      <td>0.032547823619346</td>\n",
       "      <td>0.989472879556795</td>\n",
       "      <td>0.00756174932895548</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0308388</td>\n",
       "      <td>0.986335</td>\n",
       "      <td>0.360738</td>\n",
       "      <td>0.017531</td>\n",
       "      <td>0.0177181</td>\n",
       "      <td>0.0282834</td>\n",
       "      <td>0.363502</td>\n",
       "      <td>0.0098538</td>\n",
       "      <td>0.0156143</td>\n",
       "      <td>stage i</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tcga-a3-3328</td>\n",
       "      <td>TCGA-A3-3328-01A-01D-0859-05</td>\n",
       "      <td>Beta_value</td>\n",
       "      <td>0.88549631227471</td>\n",
       "      <td>0.0384832173567634</td>\n",
       "      <td>0.0235068147586301</td>\n",
       "      <td>0.109113302762268</td>\n",
       "      <td>0.0486053663743769</td>\n",
       "      <td>0.986829026212807</td>\n",
       "      <td>0.0137776185674072</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0170604</td>\n",
       "      <td>0.989409</td>\n",
       "      <td>0.700757</td>\n",
       "      <td>0.017625</td>\n",
       "      <td>0.0189847</td>\n",
       "      <td>0.05492</td>\n",
       "      <td>0.377929</td>\n",
       "      <td>0.00975378</td>\n",
       "      <td>0.0150563</td>\n",
       "      <td>stage i</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tcga-a3-3329</td>\n",
       "      <td>TCGA-A3-3329-11A-01D-0859-05</td>\n",
       "      <td>Beta_value</td>\n",
       "      <td>0.257540543081615</td>\n",
       "      <td>0.126193684266256</td>\n",
       "      <td>0.0453059118170011</td>\n",
       "      <td>0.658352096657674</td>\n",
       "      <td>0.0417558123692779</td>\n",
       "      <td>0.987499601652852</td>\n",
       "      <td>0.0104639793481409</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0259481</td>\n",
       "      <td>0.991082</td>\n",
       "      <td>0.236762</td>\n",
       "      <td>0.0213796</td>\n",
       "      <td>0.0199632</td>\n",
       "      <td>0.046203</td>\n",
       "      <td>0.337231</td>\n",
       "      <td>0.0105036</td>\n",
       "      <td>0.017995</td>\n",
       "      <td>stage i</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21894 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  patient_barcode                       patient    category  \\\n",
       "0    tcga-a3-3325  TCGA-A3-3325-11A-01D-0859-05  Beta_value   \n",
       "1    tcga-a3-3325  TCGA-A3-3325-01A-01D-0859-05  Beta_value   \n",
       "2    tcga-a3-3328  TCGA-A3-3328-11A-01D-0859-05  Beta_value   \n",
       "3    tcga-a3-3328  TCGA-A3-3328-01A-01D-0859-05  Beta_value   \n",
       "4    tcga-a3-3329  TCGA-A3-3329-11A-01D-0859-05  Beta_value   \n",
       "\n",
       "              ATP2A1               SLMAP               MEOX2  \\\n",
       "0  0.341092896642809   0.163420373331899  0.0289863381040888   \n",
       "1    0.7253328141359   0.235892179963932   0.237244972745325   \n",
       "2   0.38595358991842   0.172516336695289  0.0267564273359283   \n",
       "3   0.88549631227471  0.0384832173567634  0.0235068147586301   \n",
       "4  0.257540543081615   0.126193684266256  0.0453059118170011   \n",
       "\n",
       "               HOXD3               PANX1     COX8C;KIAA1409  \\\n",
       "0  0.704490559626231  0.0393644321967782  0.988790175005561   \n",
       "1  0.813347696270696  0.0513241779156893  0.988732478437813   \n",
       "2  0.737048613855687   0.032547823619346  0.989472879556795   \n",
       "3  0.109113302762268  0.0486053663743769  0.986829026212807   \n",
       "4  0.658352096657674  0.0417558123692779  0.987499601652852   \n",
       "\n",
       "                  TTC8       ...           MYCL1_1 KIAA1409;COX8C ALDH1A3_1  \\\n",
       "0   0.0101987967341247       ...          0.034498       0.982875  0.280882   \n",
       "1  0.00849065030592262       ...         0.0397872        0.97624  0.471156   \n",
       "2  0.00756174932895548       ...         0.0308388       0.986335  0.360738   \n",
       "3   0.0137776185674072       ...         0.0170604       0.989409  0.700757   \n",
       "4   0.0104639793481409       ...         0.0259481       0.991082  0.236762   \n",
       "\n",
       "      ZFP2_1 C11orf24_1 LEPRE1;C1orf50_1    GNAS_1      RPN1_1    CYB5A_1  \\\n",
       "0  0.0221649  0.0195139        0.0284293  0.389761   0.0101311  0.0167906   \n",
       "1  0.0235912  0.0211279        0.0288994  0.371611   0.0100788  0.0142257   \n",
       "2   0.017531  0.0177181        0.0282834  0.363502   0.0098538  0.0156143   \n",
       "3   0.017625  0.0189847          0.05492  0.377929  0.00975378  0.0150563   \n",
       "4  0.0213796  0.0199632         0.046203  0.337231   0.0105036   0.017995   \n",
       "\n",
       "  pathologic_stage  \n",
       "0          stage i  \n",
       "1          stage i  \n",
       "2          stage i  \n",
       "3          stage i  \n",
       "4          stage i  \n",
       "\n",
       "[5 rows x 21894 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove Nan\n",
    "mixed_final = mixed_df.dropna(axis='columns',how='any') # remove column with any Nan value!\n",
    "## Detect whether there's Nan in dataframe\n",
    "mixed_final.isnull().values.any() # Detect whether there's missing value in dataframe\n",
    "mixed_final.isnull().sum().sum() # This returns an integer of the total number of NaN values\n",
    "\n",
    "## Need to add \"pathologic_stage\" back to the dataframe (it seemed to be removed during deleting Nan)!!\n",
    "mixed_final[['pathologic_stage']] = mixed_df[['pathologic_stage']] \n",
    "mixed_final.isnull().sum().sum()\n",
    "mixed_final = mixed_final.dropna(axis='index',how='any')\n",
    "mixed_final.isnull().sum().sum() \n",
    "mixed_final.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.4 Separate Tumor and Normal Samples; Label Early Stage as 1 and Label Late Stage as 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/czl/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:11: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  # This is added back by InteractiveShellApp.init_path()\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_barcode</th>\n",
       "      <th>patient</th>\n",
       "      <th>category</th>\n",
       "      <th>ATP2A1</th>\n",
       "      <th>SLMAP</th>\n",
       "      <th>MEOX2</th>\n",
       "      <th>HOXD3</th>\n",
       "      <th>PANX1</th>\n",
       "      <th>COX8C;KIAA1409</th>\n",
       "      <th>TTC8</th>\n",
       "      <th>...</th>\n",
       "      <th>KIAA1409;COX8C</th>\n",
       "      <th>ALDH1A3_1</th>\n",
       "      <th>ZFP2_1</th>\n",
       "      <th>C11orf24_1</th>\n",
       "      <th>LEPRE1;C1orf50_1</th>\n",
       "      <th>GNAS_1</th>\n",
       "      <th>RPN1_1</th>\n",
       "      <th>CYB5A_1</th>\n",
       "      <th>pathologic_stage</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tcga-a3-3325</td>\n",
       "      <td>TCGA-A3-3325-11A-01D-0859-05</td>\n",
       "      <td>Beta_value</td>\n",
       "      <td>0.341092896642809</td>\n",
       "      <td>0.163420373331899</td>\n",
       "      <td>0.0289863381040888</td>\n",
       "      <td>0.704490559626231</td>\n",
       "      <td>0.0393644321967782</td>\n",
       "      <td>0.988790175005561</td>\n",
       "      <td>0.0101987967341247</td>\n",
       "      <td>...</td>\n",
       "      <td>0.982875</td>\n",
       "      <td>0.280882</td>\n",
       "      <td>0.0221649</td>\n",
       "      <td>0.0195139</td>\n",
       "      <td>0.0284293</td>\n",
       "      <td>0.389761</td>\n",
       "      <td>0.0101311</td>\n",
       "      <td>0.0167906</td>\n",
       "      <td>stage i</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tcga-a3-3328</td>\n",
       "      <td>TCGA-A3-3328-11A-01D-0859-05</td>\n",
       "      <td>Beta_value</td>\n",
       "      <td>0.38595358991842</td>\n",
       "      <td>0.172516336695289</td>\n",
       "      <td>0.0267564273359283</td>\n",
       "      <td>0.737048613855687</td>\n",
       "      <td>0.032547823619346</td>\n",
       "      <td>0.989472879556795</td>\n",
       "      <td>0.00756174932895548</td>\n",
       "      <td>...</td>\n",
       "      <td>0.986335</td>\n",
       "      <td>0.360738</td>\n",
       "      <td>0.017531</td>\n",
       "      <td>0.0177181</td>\n",
       "      <td>0.0282834</td>\n",
       "      <td>0.363502</td>\n",
       "      <td>0.0098538</td>\n",
       "      <td>0.0156143</td>\n",
       "      <td>stage i</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tcga-a3-3329</td>\n",
       "      <td>TCGA-A3-3329-11A-01D-0859-05</td>\n",
       "      <td>Beta_value</td>\n",
       "      <td>0.257540543081615</td>\n",
       "      <td>0.126193684266256</td>\n",
       "      <td>0.0453059118170011</td>\n",
       "      <td>0.658352096657674</td>\n",
       "      <td>0.0417558123692779</td>\n",
       "      <td>0.987499601652852</td>\n",
       "      <td>0.0104639793481409</td>\n",
       "      <td>...</td>\n",
       "      <td>0.991082</td>\n",
       "      <td>0.236762</td>\n",
       "      <td>0.0213796</td>\n",
       "      <td>0.0199632</td>\n",
       "      <td>0.046203</td>\n",
       "      <td>0.337231</td>\n",
       "      <td>0.0105036</td>\n",
       "      <td>0.017995</td>\n",
       "      <td>stage i</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>tcga-a3-3336</td>\n",
       "      <td>TCGA-A3-3336-11A-01D-0859-05</td>\n",
       "      <td>Beta_value</td>\n",
       "      <td>0.401042479758341</td>\n",
       "      <td>0.154713484311169</td>\n",
       "      <td>0.0324099413145908</td>\n",
       "      <td>0.696493921392044</td>\n",
       "      <td>0.0561693907578759</td>\n",
       "      <td>0.9894664462667</td>\n",
       "      <td>0.00809906530171172</td>\n",
       "      <td>...</td>\n",
       "      <td>0.991071</td>\n",
       "      <td>0.321022</td>\n",
       "      <td>0.0191658</td>\n",
       "      <td>0.0229501</td>\n",
       "      <td>0.0315108</td>\n",
       "      <td>0.354126</td>\n",
       "      <td>0.0112805</td>\n",
       "      <td>0.0179185</td>\n",
       "      <td>stage i</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>tcga-a3-3343</td>\n",
       "      <td>TCGA-A3-3343-11A-01D-0859-05</td>\n",
       "      <td>Beta_value</td>\n",
       "      <td>0.0261269841007268</td>\n",
       "      <td>0.195132806260209</td>\n",
       "      <td>0.0442763309722378</td>\n",
       "      <td>0.271151912048597</td>\n",
       "      <td>0.0895605935074786</td>\n",
       "      <td>0.985994929771935</td>\n",
       "      <td>0.0111629863693614</td>\n",
       "      <td>...</td>\n",
       "      <td>0.991888</td>\n",
       "      <td>0.227373</td>\n",
       "      <td>0.0361556</td>\n",
       "      <td>0.0309748</td>\n",
       "      <td>0.0450911</td>\n",
       "      <td>0.21732</td>\n",
       "      <td>0.00884356</td>\n",
       "      <td>0.0268121</td>\n",
       "      <td>stage ii</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21895 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  patient_barcode                       patient    category  \\\n",
       "0    tcga-a3-3325  TCGA-A3-3325-11A-01D-0859-05  Beta_value   \n",
       "2    tcga-a3-3328  TCGA-A3-3328-11A-01D-0859-05  Beta_value   \n",
       "4    tcga-a3-3329  TCGA-A3-3329-11A-01D-0859-05  Beta_value   \n",
       "6    tcga-a3-3336  TCGA-A3-3336-11A-01D-0859-05  Beta_value   \n",
       "8    tcga-a3-3343  TCGA-A3-3343-11A-01D-0859-05  Beta_value   \n",
       "\n",
       "               ATP2A1              SLMAP               MEOX2  \\\n",
       "0   0.341092896642809  0.163420373331899  0.0289863381040888   \n",
       "2    0.38595358991842  0.172516336695289  0.0267564273359283   \n",
       "4   0.257540543081615  0.126193684266256  0.0453059118170011   \n",
       "6   0.401042479758341  0.154713484311169  0.0324099413145908   \n",
       "8  0.0261269841007268  0.195132806260209  0.0442763309722378   \n",
       "\n",
       "               HOXD3               PANX1     COX8C;KIAA1409  \\\n",
       "0  0.704490559626231  0.0393644321967782  0.988790175005561   \n",
       "2  0.737048613855687   0.032547823619346  0.989472879556795   \n",
       "4  0.658352096657674  0.0417558123692779  0.987499601652852   \n",
       "6  0.696493921392044  0.0561693907578759    0.9894664462667   \n",
       "8  0.271151912048597  0.0895605935074786  0.985994929771935   \n",
       "\n",
       "                  TTC8  ...  KIAA1409;COX8C ALDH1A3_1     ZFP2_1 C11orf24_1  \\\n",
       "0   0.0101987967341247  ...        0.982875  0.280882  0.0221649  0.0195139   \n",
       "2  0.00756174932895548  ...        0.986335  0.360738   0.017531  0.0177181   \n",
       "4   0.0104639793481409  ...        0.991082  0.236762  0.0213796  0.0199632   \n",
       "6  0.00809906530171172  ...        0.991071  0.321022  0.0191658  0.0229501   \n",
       "8   0.0111629863693614  ...        0.991888  0.227373  0.0361556  0.0309748   \n",
       "\n",
       "  LEPRE1;C1orf50_1    GNAS_1      RPN1_1    CYB5A_1 pathologic_stage group  \n",
       "0        0.0284293  0.389761   0.0101311  0.0167906          stage i     1  \n",
       "2        0.0282834  0.363502   0.0098538  0.0156143          stage i     1  \n",
       "4         0.046203  0.337231   0.0105036   0.017995          stage i     1  \n",
       "6        0.0315108  0.354126   0.0112805  0.0179185          stage i     1  \n",
       "8        0.0450911   0.21732  0.00884356  0.0268121         stage ii     1  \n",
       "\n",
       "[5 rows x 21895 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Separate tumor tissue with normal tissue\n",
    "tumor_list = []\n",
    "for n in range(int(mixed_final.shape[0]/2)):\n",
    "    m = 2*n\n",
    "    tumor_list.append(m)\n",
    "\n",
    "tumor_df = mixed_final.iloc[tumor_list,:]\n",
    "\n",
    "## Label tumor stage 1,2 to be group 1; tumor stage 3,4  to be group 2\n",
    "dic = {'stage i':1, 'stage ii':1, 'stage iii':2, 'stage iv':2}\n",
    "tumor_df['group']=tumor_df['pathologic_stage'].map(dic)\n",
    "tumor_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.5 Export Clean Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ATP2A1</th>\n",
       "      <th>SLMAP</th>\n",
       "      <th>MEOX2</th>\n",
       "      <th>HOXD3</th>\n",
       "      <th>PANX1</th>\n",
       "      <th>COX8C;KIAA1409</th>\n",
       "      <th>TTC8</th>\n",
       "      <th>nan</th>\n",
       "      <th>TMEM186;PMM2</th>\n",
       "      <th>ANG;RNASE4</th>\n",
       "      <th>...</th>\n",
       "      <th>MYCL1_1</th>\n",
       "      <th>KIAA1409;COX8C</th>\n",
       "      <th>ALDH1A3_1</th>\n",
       "      <th>ZFP2_1</th>\n",
       "      <th>C11orf24_1</th>\n",
       "      <th>LEPRE1;C1orf50_1</th>\n",
       "      <th>GNAS_1</th>\n",
       "      <th>RPN1_1</th>\n",
       "      <th>CYB5A_1</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.341092896642809</td>\n",
       "      <td>0.163420373331899</td>\n",
       "      <td>0.0289863381040888</td>\n",
       "      <td>0.704490559626231</td>\n",
       "      <td>0.0393644321967782</td>\n",
       "      <td>0.988790175005561</td>\n",
       "      <td>0.0101987967341247</td>\n",
       "      <td>0.553604734439809</td>\n",
       "      <td>0.953507442534218</td>\n",
       "      <td>0.0145064421302908</td>\n",
       "      <td>...</td>\n",
       "      <td>0.034498</td>\n",
       "      <td>0.982875</td>\n",
       "      <td>0.280882</td>\n",
       "      <td>0.0221649</td>\n",
       "      <td>0.0195139</td>\n",
       "      <td>0.0284293</td>\n",
       "      <td>0.389761</td>\n",
       "      <td>0.0101311</td>\n",
       "      <td>0.0167906</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.38595358991842</td>\n",
       "      <td>0.172516336695289</td>\n",
       "      <td>0.0267564273359283</td>\n",
       "      <td>0.737048613855687</td>\n",
       "      <td>0.032547823619346</td>\n",
       "      <td>0.989472879556795</td>\n",
       "      <td>0.00756174932895548</td>\n",
       "      <td>0.620752883476738</td>\n",
       "      <td>0.953492118622301</td>\n",
       "      <td>0.0104582276294671</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0308388</td>\n",
       "      <td>0.986335</td>\n",
       "      <td>0.360738</td>\n",
       "      <td>0.017531</td>\n",
       "      <td>0.0177181</td>\n",
       "      <td>0.0282834</td>\n",
       "      <td>0.363502</td>\n",
       "      <td>0.0098538</td>\n",
       "      <td>0.0156143</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.257540543081615</td>\n",
       "      <td>0.126193684266256</td>\n",
       "      <td>0.0453059118170011</td>\n",
       "      <td>0.658352096657674</td>\n",
       "      <td>0.0417558123692779</td>\n",
       "      <td>0.987499601652852</td>\n",
       "      <td>0.0104639793481409</td>\n",
       "      <td>0.652652809982813</td>\n",
       "      <td>0.914208531588916</td>\n",
       "      <td>0.0238009339117171</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0259481</td>\n",
       "      <td>0.991082</td>\n",
       "      <td>0.236762</td>\n",
       "      <td>0.0213796</td>\n",
       "      <td>0.0199632</td>\n",
       "      <td>0.046203</td>\n",
       "      <td>0.337231</td>\n",
       "      <td>0.0105036</td>\n",
       "      <td>0.017995</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.401042479758341</td>\n",
       "      <td>0.154713484311169</td>\n",
       "      <td>0.0324099413145908</td>\n",
       "      <td>0.696493921392044</td>\n",
       "      <td>0.0561693907578759</td>\n",
       "      <td>0.9894664462667</td>\n",
       "      <td>0.00809906530171172</td>\n",
       "      <td>0.615327033621039</td>\n",
       "      <td>0.949000471552509</td>\n",
       "      <td>0.0126582340654666</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0427957</td>\n",
       "      <td>0.991071</td>\n",
       "      <td>0.321022</td>\n",
       "      <td>0.0191658</td>\n",
       "      <td>0.0229501</td>\n",
       "      <td>0.0315108</td>\n",
       "      <td>0.354126</td>\n",
       "      <td>0.0112805</td>\n",
       "      <td>0.0179185</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.0261269841007268</td>\n",
       "      <td>0.195132806260209</td>\n",
       "      <td>0.0442763309722378</td>\n",
       "      <td>0.271151912048597</td>\n",
       "      <td>0.0895605935074786</td>\n",
       "      <td>0.985994929771935</td>\n",
       "      <td>0.0111629863693614</td>\n",
       "      <td>0.509603947268456</td>\n",
       "      <td>0.926462406994679</td>\n",
       "      <td>0.0213652841338812</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0533251</td>\n",
       "      <td>0.991888</td>\n",
       "      <td>0.227373</td>\n",
       "      <td>0.0361556</td>\n",
       "      <td>0.0309748</td>\n",
       "      <td>0.0450911</td>\n",
       "      <td>0.21732</td>\n",
       "      <td>0.00884356</td>\n",
       "      <td>0.0268121</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21891 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               ATP2A1              SLMAP               MEOX2  \\\n",
       "0   0.341092896642809  0.163420373331899  0.0289863381040888   \n",
       "2    0.38595358991842  0.172516336695289  0.0267564273359283   \n",
       "4   0.257540543081615  0.126193684266256  0.0453059118170011   \n",
       "6   0.401042479758341  0.154713484311169  0.0324099413145908   \n",
       "8  0.0261269841007268  0.195132806260209  0.0442763309722378   \n",
       "\n",
       "               HOXD3               PANX1     COX8C;KIAA1409  \\\n",
       "0  0.704490559626231  0.0393644321967782  0.988790175005561   \n",
       "2  0.737048613855687   0.032547823619346  0.989472879556795   \n",
       "4  0.658352096657674  0.0417558123692779  0.987499601652852   \n",
       "6  0.696493921392044  0.0561693907578759    0.9894664462667   \n",
       "8  0.271151912048597  0.0895605935074786  0.985994929771935   \n",
       "\n",
       "                  TTC8                NaN       TMEM186;PMM2  \\\n",
       "0   0.0101987967341247  0.553604734439809  0.953507442534218   \n",
       "2  0.00756174932895548  0.620752883476738  0.953492118622301   \n",
       "4   0.0104639793481409  0.652652809982813  0.914208531588916   \n",
       "6  0.00809906530171172  0.615327033621039  0.949000471552509   \n",
       "8   0.0111629863693614  0.509603947268456  0.926462406994679   \n",
       "\n",
       "           ANG;RNASE4  ...     MYCL1_1 KIAA1409;COX8C ALDH1A3_1     ZFP2_1  \\\n",
       "0  0.0145064421302908  ...    0.034498       0.982875  0.280882  0.0221649   \n",
       "2  0.0104582276294671  ...   0.0308388       0.986335  0.360738   0.017531   \n",
       "4  0.0238009339117171  ...   0.0259481       0.991082  0.236762  0.0213796   \n",
       "6  0.0126582340654666  ...   0.0427957       0.991071  0.321022  0.0191658   \n",
       "8  0.0213652841338812  ...   0.0533251       0.991888  0.227373  0.0361556   \n",
       "\n",
       "  C11orf24_1 LEPRE1;C1orf50_1    GNAS_1      RPN1_1    CYB5A_1 group  \n",
       "0  0.0195139        0.0284293  0.389761   0.0101311  0.0167906     1  \n",
       "2  0.0177181        0.0282834  0.363502   0.0098538  0.0156143     1  \n",
       "4  0.0199632         0.046203  0.337231   0.0105036   0.017995     1  \n",
       "6  0.0229501        0.0315108  0.354126   0.0112805  0.0179185     1  \n",
       "8  0.0309748        0.0450911   0.21732  0.00884356  0.0268121     1  \n",
       "\n",
       "[5 rows x 21891 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Only keep genes methylation and label columns\n",
    "## Export clean dataset\n",
    "X_tumor = tumor_df.drop(['patient_barcode', 'pathologic_stage', 'patient', 'category'],axis=1)\n",
    "X_tumor.to_csv('Tumor_stage.csv')\n",
    "X_tumor.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 2. Feature Selection & Extraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ATP2A1</th>\n",
       "      <th>SLMAP</th>\n",
       "      <th>MEOX2</th>\n",
       "      <th>HOXD3</th>\n",
       "      <th>PANX1</th>\n",
       "      <th>COX8C;KIAA1409</th>\n",
       "      <th>TTC8</th>\n",
       "      <th>Unnamed: 8</th>\n",
       "      <th>TMEM186;PMM2</th>\n",
       "      <th>ANG;RNASE4</th>\n",
       "      <th>...</th>\n",
       "      <th>MYCL1_1</th>\n",
       "      <th>KIAA1409;COX8C</th>\n",
       "      <th>ALDH1A3_1.4</th>\n",
       "      <th>ZFP2_1</th>\n",
       "      <th>C11orf24_1</th>\n",
       "      <th>LEPRE1;C1orf50_1.1</th>\n",
       "      <th>GNAS_1.15</th>\n",
       "      <th>RPN1_1</th>\n",
       "      <th>CYB5A_1</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.341093</td>\n",
       "      <td>0.163420</td>\n",
       "      <td>0.028986</td>\n",
       "      <td>0.704491</td>\n",
       "      <td>0.039364</td>\n",
       "      <td>0.988790</td>\n",
       "      <td>0.010199</td>\n",
       "      <td>0.553605</td>\n",
       "      <td>0.953507</td>\n",
       "      <td>0.014506</td>\n",
       "      <td>...</td>\n",
       "      <td>0.034498</td>\n",
       "      <td>0.982875</td>\n",
       "      <td>0.280882</td>\n",
       "      <td>0.022165</td>\n",
       "      <td>0.019514</td>\n",
       "      <td>0.028429</td>\n",
       "      <td>0.389761</td>\n",
       "      <td>0.010131</td>\n",
       "      <td>0.016791</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.385954</td>\n",
       "      <td>0.172516</td>\n",
       "      <td>0.026756</td>\n",
       "      <td>0.737049</td>\n",
       "      <td>0.032548</td>\n",
       "      <td>0.989473</td>\n",
       "      <td>0.007562</td>\n",
       "      <td>0.620753</td>\n",
       "      <td>0.953492</td>\n",
       "      <td>0.010458</td>\n",
       "      <td>...</td>\n",
       "      <td>0.030839</td>\n",
       "      <td>0.986335</td>\n",
       "      <td>0.360738</td>\n",
       "      <td>0.017531</td>\n",
       "      <td>0.017718</td>\n",
       "      <td>0.028283</td>\n",
       "      <td>0.363502</td>\n",
       "      <td>0.009854</td>\n",
       "      <td>0.015614</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.257541</td>\n",
       "      <td>0.126194</td>\n",
       "      <td>0.045306</td>\n",
       "      <td>0.658352</td>\n",
       "      <td>0.041756</td>\n",
       "      <td>0.987500</td>\n",
       "      <td>0.010464</td>\n",
       "      <td>0.652653</td>\n",
       "      <td>0.914209</td>\n",
       "      <td>0.023801</td>\n",
       "      <td>...</td>\n",
       "      <td>0.025948</td>\n",
       "      <td>0.991082</td>\n",
       "      <td>0.236762</td>\n",
       "      <td>0.021380</td>\n",
       "      <td>0.019963</td>\n",
       "      <td>0.046203</td>\n",
       "      <td>0.337231</td>\n",
       "      <td>0.010504</td>\n",
       "      <td>0.017995</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.401042</td>\n",
       "      <td>0.154713</td>\n",
       "      <td>0.032410</td>\n",
       "      <td>0.696494</td>\n",
       "      <td>0.056169</td>\n",
       "      <td>0.989466</td>\n",
       "      <td>0.008099</td>\n",
       "      <td>0.615327</td>\n",
       "      <td>0.949000</td>\n",
       "      <td>0.012658</td>\n",
       "      <td>...</td>\n",
       "      <td>0.042796</td>\n",
       "      <td>0.991071</td>\n",
       "      <td>0.321022</td>\n",
       "      <td>0.019166</td>\n",
       "      <td>0.022950</td>\n",
       "      <td>0.031511</td>\n",
       "      <td>0.354126</td>\n",
       "      <td>0.011281</td>\n",
       "      <td>0.017918</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.026127</td>\n",
       "      <td>0.195133</td>\n",
       "      <td>0.044276</td>\n",
       "      <td>0.271152</td>\n",
       "      <td>0.089561</td>\n",
       "      <td>0.985995</td>\n",
       "      <td>0.011163</td>\n",
       "      <td>0.509604</td>\n",
       "      <td>0.926462</td>\n",
       "      <td>0.021365</td>\n",
       "      <td>...</td>\n",
       "      <td>0.053325</td>\n",
       "      <td>0.991888</td>\n",
       "      <td>0.227373</td>\n",
       "      <td>0.036156</td>\n",
       "      <td>0.030975</td>\n",
       "      <td>0.045091</td>\n",
       "      <td>0.217320</td>\n",
       "      <td>0.008844</td>\n",
       "      <td>0.026812</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21891 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ATP2A1     SLMAP     MEOX2     HOXD3     PANX1  COX8C;KIAA1409      TTC8  \\\n",
       "0  0.341093  0.163420  0.028986  0.704491  0.039364        0.988790  0.010199   \n",
       "1  0.385954  0.172516  0.026756  0.737049  0.032548        0.989473  0.007562   \n",
       "2  0.257541  0.126194  0.045306  0.658352  0.041756        0.987500  0.010464   \n",
       "3  0.401042  0.154713  0.032410  0.696494  0.056169        0.989466  0.008099   \n",
       "4  0.026127  0.195133  0.044276  0.271152  0.089561        0.985995  0.011163   \n",
       "\n",
       "   Unnamed: 8  TMEM186;PMM2  ANG;RNASE4  ...     MYCL1_1  KIAA1409;COX8C  \\\n",
       "0    0.553605      0.953507    0.014506  ...    0.034498        0.982875   \n",
       "1    0.620753      0.953492    0.010458  ...    0.030839        0.986335   \n",
       "2    0.652653      0.914209    0.023801  ...    0.025948        0.991082   \n",
       "3    0.615327      0.949000    0.012658  ...    0.042796        0.991071   \n",
       "4    0.509604      0.926462    0.021365  ...    0.053325        0.991888   \n",
       "\n",
       "   ALDH1A3_1.4    ZFP2_1  C11orf24_1  LEPRE1;C1orf50_1.1  GNAS_1.15    RPN1_1  \\\n",
       "0     0.280882  0.022165    0.019514            0.028429   0.389761  0.010131   \n",
       "1     0.360738  0.017531    0.017718            0.028283   0.363502  0.009854   \n",
       "2     0.236762  0.021380    0.019963            0.046203   0.337231  0.010504   \n",
       "3     0.321022  0.019166    0.022950            0.031511   0.354126  0.011281   \n",
       "4     0.227373  0.036156    0.030975            0.045091   0.217320  0.008844   \n",
       "\n",
       "    CYB5A_1  group  \n",
       "0  0.016791      1  \n",
       "1  0.015614      1  \n",
       "2  0.017995      1  \n",
       "3  0.017918      1  \n",
       "4  0.026812      1  \n",
       "\n",
       "[5 rows x 21891 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Data Loading 1 (Total data)\n",
    "## From above clean dataset\n",
    "X_tum = pd.read_csv('Tumor_stage.csv')\n",
    "X_tumor = X_tum.drop(['Unnamed: 0'], axis=1)\n",
    "y_tumor = X_tum[['group']]\n",
    "X_tumor.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.1 Train-Test Split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ATP2A1</th>\n",
       "      <th>SLMAP</th>\n",
       "      <th>MEOX2</th>\n",
       "      <th>HOXD3</th>\n",
       "      <th>PANX1</th>\n",
       "      <th>COX8C;KIAA1409</th>\n",
       "      <th>TTC8</th>\n",
       "      <th>Unnamed: 8</th>\n",
       "      <th>TMEM186;PMM2</th>\n",
       "      <th>ANG;RNASE4</th>\n",
       "      <th>...</th>\n",
       "      <th>DAB2IP_1.6</th>\n",
       "      <th>MYCL1_1</th>\n",
       "      <th>KIAA1409;COX8C</th>\n",
       "      <th>ALDH1A3_1.4</th>\n",
       "      <th>ZFP2_1</th>\n",
       "      <th>C11orf24_1</th>\n",
       "      <th>LEPRE1;C1orf50_1.1</th>\n",
       "      <th>GNAS_1.15</th>\n",
       "      <th>RPN1_1</th>\n",
       "      <th>CYB5A_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>0.527216</td>\n",
       "      <td>0.388513</td>\n",
       "      <td>0.041617</td>\n",
       "      <td>0.821631</td>\n",
       "      <td>0.082274</td>\n",
       "      <td>0.990079</td>\n",
       "      <td>0.009101</td>\n",
       "      <td>0.595033</td>\n",
       "      <td>0.948046</td>\n",
       "      <td>0.012949</td>\n",
       "      <td>...</td>\n",
       "      <td>0.014801</td>\n",
       "      <td>0.261126</td>\n",
       "      <td>0.967535</td>\n",
       "      <td>0.603313</td>\n",
       "      <td>0.017250</td>\n",
       "      <td>0.025930</td>\n",
       "      <td>0.040611</td>\n",
       "      <td>0.355288</td>\n",
       "      <td>0.012419</td>\n",
       "      <td>0.029987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>0.423040</td>\n",
       "      <td>0.148296</td>\n",
       "      <td>0.029480</td>\n",
       "      <td>0.780460</td>\n",
       "      <td>0.076754</td>\n",
       "      <td>0.990518</td>\n",
       "      <td>0.009348</td>\n",
       "      <td>0.633822</td>\n",
       "      <td>0.910637</td>\n",
       "      <td>0.014222</td>\n",
       "      <td>...</td>\n",
       "      <td>0.246410</td>\n",
       "      <td>0.059515</td>\n",
       "      <td>0.977434</td>\n",
       "      <td>0.192991</td>\n",
       "      <td>0.016769</td>\n",
       "      <td>0.027705</td>\n",
       "      <td>0.052330</td>\n",
       "      <td>0.303668</td>\n",
       "      <td>0.009113</td>\n",
       "      <td>0.024134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>0.391904</td>\n",
       "      <td>0.354476</td>\n",
       "      <td>0.053835</td>\n",
       "      <td>0.794001</td>\n",
       "      <td>0.056835</td>\n",
       "      <td>0.991049</td>\n",
       "      <td>0.010656</td>\n",
       "      <td>0.576181</td>\n",
       "      <td>0.918385</td>\n",
       "      <td>0.013271</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007594</td>\n",
       "      <td>0.184651</td>\n",
       "      <td>0.969820</td>\n",
       "      <td>0.445231</td>\n",
       "      <td>0.032726</td>\n",
       "      <td>0.022425</td>\n",
       "      <td>0.049316</td>\n",
       "      <td>0.338151</td>\n",
       "      <td>0.009884</td>\n",
       "      <td>0.031497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>0.544094</td>\n",
       "      <td>0.291288</td>\n",
       "      <td>0.068571</td>\n",
       "      <td>0.758822</td>\n",
       "      <td>0.046573</td>\n",
       "      <td>0.990030</td>\n",
       "      <td>0.010247</td>\n",
       "      <td>0.704257</td>\n",
       "      <td>0.928793</td>\n",
       "      <td>0.015198</td>\n",
       "      <td>...</td>\n",
       "      <td>0.147100</td>\n",
       "      <td>0.027463</td>\n",
       "      <td>0.988602</td>\n",
       "      <td>0.287220</td>\n",
       "      <td>0.032016</td>\n",
       "      <td>0.030268</td>\n",
       "      <td>0.070115</td>\n",
       "      <td>0.402414</td>\n",
       "      <td>0.011178</td>\n",
       "      <td>0.032123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>0.382834</td>\n",
       "      <td>0.281154</td>\n",
       "      <td>0.054028</td>\n",
       "      <td>0.776502</td>\n",
       "      <td>0.051761</td>\n",
       "      <td>0.990223</td>\n",
       "      <td>0.010292</td>\n",
       "      <td>0.718880</td>\n",
       "      <td>0.933810</td>\n",
       "      <td>0.013489</td>\n",
       "      <td>...</td>\n",
       "      <td>0.138122</td>\n",
       "      <td>0.050770</td>\n",
       "      <td>0.992555</td>\n",
       "      <td>0.241898</td>\n",
       "      <td>0.038682</td>\n",
       "      <td>0.034609</td>\n",
       "      <td>0.049532</td>\n",
       "      <td>0.450248</td>\n",
       "      <td>0.013350</td>\n",
       "      <td>0.021261</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21890 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       ATP2A1     SLMAP     MEOX2     HOXD3     PANX1  COX8C;KIAA1409  \\\n",
       "25   0.527216  0.388513  0.041617  0.821631  0.082274        0.990079   \n",
       "124  0.423040  0.148296  0.029480  0.780460  0.076754        0.990518   \n",
       "12   0.391904  0.354476  0.053835  0.794001  0.056835        0.991049   \n",
       "185  0.544094  0.291288  0.068571  0.758822  0.046573        0.990030   \n",
       "140  0.382834  0.281154  0.054028  0.776502  0.051761        0.990223   \n",
       "\n",
       "         TTC8  Unnamed: 8  TMEM186;PMM2  ANG;RNASE4    ...     DAB2IP_1.6  \\\n",
       "25   0.009101    0.595033      0.948046    0.012949    ...       0.014801   \n",
       "124  0.009348    0.633822      0.910637    0.014222    ...       0.246410   \n",
       "12   0.010656    0.576181      0.918385    0.013271    ...       0.007594   \n",
       "185  0.010247    0.704257      0.928793    0.015198    ...       0.147100   \n",
       "140  0.010292    0.718880      0.933810    0.013489    ...       0.138122   \n",
       "\n",
       "      MYCL1_1  KIAA1409;COX8C  ALDH1A3_1.4    ZFP2_1  C11orf24_1  \\\n",
       "25   0.261126        0.967535     0.603313  0.017250    0.025930   \n",
       "124  0.059515        0.977434     0.192991  0.016769    0.027705   \n",
       "12   0.184651        0.969820     0.445231  0.032726    0.022425   \n",
       "185  0.027463        0.988602     0.287220  0.032016    0.030268   \n",
       "140  0.050770        0.992555     0.241898  0.038682    0.034609   \n",
       "\n",
       "     LEPRE1;C1orf50_1.1  GNAS_1.15    RPN1_1   CYB5A_1  \n",
       "25             0.040611   0.355288  0.012419  0.029987  \n",
       "124            0.052330   0.303668  0.009113  0.024134  \n",
       "12             0.049316   0.338151  0.009884  0.031497  \n",
       "185            0.070115   0.402414  0.011178  0.032123  \n",
       "140            0.049532   0.450248  0.013350  0.021261  \n",
       "\n",
       "[5 rows x 21890 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn.model_selection import train_test_split # https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html\n",
    "\n",
    "X_train_di, X_test_di, y_train_di, y_test_di = train_test_split (X_tumor, y_tumor, test_size = 0.3, stratify=X_tumor['group'])\n",
    "X_train_dif = X_train_di.drop(['group'],axis=1)\n",
    "X_test_dif = X_test_di.drop(['group'],axis=1)\n",
    "\n",
    "X_train_dif.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.2 Feature Selection: Kruskal-Wallis Method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "95\n",
      "20\n",
      "6\n",
      "3659\n"
     ]
    }
   ],
   "source": [
    "from scipy import stats\n",
    "## Split Early and Late Stage Samples; Only use the train test sample!\n",
    "tumor_G1 = X_tumor[X_tumor.group == 1].drop(['group'],axis=1)\n",
    "tumor_G2 = X_tumor[X_tumor.group == 2].drop(['group'],axis=1)\n",
    "Gene = tumor_G1.columns.tolist() # Have a list of genes\n",
    "\n",
    "## For multiple pair-comparison, type I error keep accumulating! So we need higher threshold\n",
    "## Setting Different threshold for p < alpha(type I error): alpha = 5e-4, 5e-5, 1e-5\n",
    "alpha = [5e-4, 5e-5, 1e-5, 5e-2]\n",
    "gene_kw = [[],[],[],[]]\n",
    "for n in range(len(alpha)):   \n",
    "    for z in range(len(Gene)):\n",
    "        G1 = tumor_G1.iloc[:,z]\n",
    "        G2 = tumor_G2.iloc[:,z]\n",
    "        p = stats.kruskal(G1, G2)[1]\n",
    "        \n",
    "        if p < alpha[n]:\n",
    "            gene_kw[n].append(Gene[z])\n",
    "    print(len(gene_kw[n]))    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.3 Feature Selection: LDA (Linear Discriminant Analysis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/czl/anaconda3/lib/python3.7/site-packages/sklearn/utils/validation.py:761: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().\n",
      "  y = column_or_1d(y, warn=True)\n",
      "/home/czl/anaconda3/lib/python3.7/site-packages/sklearn/discriminant_analysis.py:388: UserWarning: Variables are collinear.\n",
      "  warnings.warn(\"Variables are collinear.\")\n"
     ]
    }
   ],
   "source": [
    "from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA\n",
    "     # https://scikit-learn.org/stable/modules/generated/sklearn.discriminant_analysis.LinearDiscriminantAnalysis.html\n",
    "# Linear Discriminant Analysis (LDA):\n",
    "  # A classifier with a linear decision boundary, generated by fitting class conditional densities to the data and using Bayes’ rule\n",
    "  # The model fits a Gaussian density to each class, assuming that all classes share the same covariance matrix.  \n",
    "\n",
    "lda = LDA(solver=\"svd\", store_covariance=True) #Singular value decomposition (default). Does not compute the covariance matrix, therefore this solver is recommended for data with a large number of features.\n",
    "Tumor = X_tumor.drop(['group'], axis=1)\n",
    "X = Tumor[gene_kw[3]] #  whole data is too large for LDA!\n",
    "y = y_tumor\n",
    "\n",
    "lda.fit(X,y)\n",
    "coef = lda.coef_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "95\n",
      "20\n",
      "6\n"
     ]
    }
   ],
   "source": [
    "## 1. Sort the coefficients, then select top 95, 20, and 6 features\n",
    "## 2. Save index, then select corresponding genes!\n",
    "g_coef = coef[0]\n",
    "number = [95,20,6]\n",
    "gene_lda = [[],[],[]]\n",
    "\n",
    "for n in range(len(number)):\n",
    "    index = np.argsort(g_coef)[:number[n]] # numpy.argsort(list): sort the list and output the original index\n",
    "    gene_lda[n] = [Gene[i] for i in index]\n",
    "    print(len(gene_lda[n]))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
